File: conversions/ios/swiftui/ConversionsSwiftUI/DeepLink.swift
//
//  DeepLink.swift
//  DubSwiftUIDemo
//
//  Created by Ian MacCallum on 9/18/25.
//

import Foundation

enum DeepLink {
    case product(id: String)

    init?(from urlString: String) {
        guard let url = URL(string: urlString) else {
            return nil
        }

        let pathComponents = url.pathComponents
        let productId: String?

        if let productIndex = pathComponents.firstIndex(of: "product"),
           productIndex + 1 < pathComponents.count {
            productId = pathComponents[productIndex + 1]
        } else {
            productId = pathComponents.last
        }

        guard let productId = productId, !productId.isEmpty else {
            return nil
        }

        self = .product(id: productId)
    }
}

File: conversions/react-native/react-navigation/src/lib/deep-link.ts
export interface DeepLink {
  type: 'product';
  id: string;
}

export const parseDeepLink = (urlString: string): DeepLink | null => {
  try {
    const url = new URL(urlString);
    const pathComponents = url.pathname.split('/').filter(Boolean);

    let productId: string | null = null;

    const productIndex = pathComponents.findIndex(
      (component) => component === 'product'
    );

    if (productIndex !== -1 && productIndex + 1 < pathComponents.length) {
      productId = pathComponents[productIndex + 1] || null;
    } else {
      productId = pathComponents[pathComponents.length - 1] || null;
    }

    if (productId && productId.length > 0) {
      return {
        type: 'product',
        id: productId,
      };
    }

    return null;
  } catch (error) {
    console.error('Error parsing deep link:', error);
    return null;
  }
};
